keras/src/distribution/tensor_parallel/autoconfig.py

```diff
@@ -0,0 +1,262 @@
+        full_name: The full hierarchical name of the layer (prefix + name).
+        device_count: Total number of devices.
+        state_rules: The dictionary to update with variable sharding rules.
+        output_rules: The dictionary to update with output layout rules.
+    """
+    if isinstance(layer, layers.Dense):
+        mlp_type = analyze_dense_layer(layer)
+
+        if mlp_type == "up_projection":
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=1
+            )
+            if layer.use_bias:
+                state_rules[f"{full_name}.bias"] = _split_rule(
+                    device_count, dim=0
+                )
+            output_rules[f"{full_name}"] = {0: "gather"}
+
+        elif mlp_type == "down_projection":
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=0
+            )
+            output_rules[f"{full_name}"] = {0: "allreduce"}
+
+        else:
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=1
+            )
+            if layer.use_bias:
+                state_rules[f"{full_name}.bias"] = _split_rule(
+                    device_count, dim=0
+                )
+            output_rules[f"{full_name}"] = {0: "gather -1"}
+
+    elif isinstance(layer, layers.EinsumDense):
+        if "attention_output" in full_name:
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=0
+            )
+            output_rules[f"{full_name}"] = {0: "allreduce"}
+        else:
+            state_rules[f"{full_name}.kernel"] = _split_rule(
+                device_count, dim=1
+            )
+            if hasattr(layer, "bias") and layer.bias is not None:
+                state_rules[f"{full_name}.bias"] = _split_rule(
+                    device_count, dim=0
+                )
+            output_rules[f"{full_name}"] = {0: "gather -1"}
+
+    elif (
+        isinstance(layer, (layers.Embedding,))
+        or "Embedding" in layer.__class__.__name__
+    ):
+        if hasattr(layer, "weights"):
+            for weight in layer.weights:
+                if "embedding" in weight.name or "weight" in weight.name:
+                    key_found = False
+                    for attr_candidate in [
+                        "embeddings",
+                        "position_embeddings",
+                        "weight",
+                    ]:
+                        if getattr(layer, attr_candidate, None) is weight:
+                            state_rules[f"{full_name}.{attr_candidate}"] = (
+                                _split_rule(device_count, dim=1)
+                            )
+                            key_found = True
+                            break
+
+                    if not key_found:
+                        clean_name = weight.name.split("/")[-1].split(":")[0]
```

Comment:
When do variable names contain ":"?

To me this was a legacy thing in Keras 2 that we don't have in Keras 3.

What about "/"? It should be in the path, not the name.